tools/nni_trial_tool/trial_keeper.py

```diff
@@ -90,13 +92,13 @@ def main_loop(args):
 
     # Notice: We don't appoint env, which means subprocess wil inherit current environment and that is expected behavior
     log_pipe_stdout = trial_syslogger_stdout.get_pipelog_reader()
-    process = Popen(args.trial_command, shell=True, stdout=log_pipe_stdout, stderr=log_pipe_stdout)
-    nni_log(LogType.Info, 'Trial keeper spawns a subprocess (pid {0}) to run command: {1}'.format(process.pid,
+    _trial_process = Popen(args.trial_command, shell=True, stdout=log_pipe_stdout, stderr=log_pipe_stdout, preexec_fn=os.setsid)
```

Comment:
what is `preexec_fn=os.setsid` used for?

Reply:
https://github.com/microsoft/nni/blob/95bb20be17adda7375128c306bfa58a8ac6e31b8/tools/nni_trial_tool/trial_keeper.py#L225-L226

in order to properly synchronize and kill the process in deamon with pid:
* https://stackoverflow.com/a/22582602 a practice in Python ([example1](https://github.com/google/google-ctf/blob/bce4a1410d910af65e725fa6896aee0718e961c2/third_party/edk2/AppPkg/Applications/Python/Python-2.7.2/Lib/webbrowser.py#L198), [example2](https://github.com/facebook/FAI-PEP/blob/f4274b85e4a74faf905071262cce11bae1314c8c/benchmarking/utils/subprocess_with_logger.py#L186))
* https://stackoverflow.com/a/63000117 pid & sid, a recent answer